### ISSUE TITLE
Thread-local variants 

### ISSUE THREAD

**Issue by re-xyr** [re-xyr/cleff#17]
Although the `State` and `Writer` effects are shared by default (due to Haskell having shared mutvars only), it will be nice if we have thread-local variants too.

This can be implemented with something like `thread-util-context`:

```haskell
handleStateLocal :: s -> Eff (State s : es) ~> Eff es
handleStateLocal s0 m = thisIsPureTrustMe do
  tsm <- newThreadStorageMap
  reinterpret (handle tsm) m
  where
    handle tsm = \case
      Get -> lookup tsm >>= \case
        Nothing -> attach tsm s0 $> s0
        Just s -> s
      Put s' -> void $ attach tsm s'
```

(`State` omitted. It cannot be atomic via the `thread-utils-context` API, but is implementable.)

**Comment by re-xyr** [comment]
Note that this is *different* from the `effectful` semantics:

- In `effectful`, forked threads inherits the value of the forker thread
- Here, forked threads have a fixed initial value passed to the handler

Both semantics have merits, but the former cannot be implemented without providing a custom `forkIO` (as well as the async ecosystem built on it), which is something I don't want to do.

**Comment by re-xyr** [comment]
Turns out this implementation won't support pooled threads... bummers. `effectful`'s way of solving this (hooks in `withRunInIO`) has some pretty flaky assumptions, so while they do work with `unliftio`'s pooled functions, they may not work with other implementations. I'd rather leave things as is.

**Comment by re-xyr** [comment]
Solved in 5668dc3
